src/AutoRest.CSharp/Common/Output/Models/Types/ModelFactoryTypeProvider.cs

```diff
@@ -29,26 +31,39 @@ internal sealed class ModelFactoryTypeProvider : TypeProvider
         protected override string DefaultName { get; }
         protected override string DefaultAccessibility { get; }
 
-        private IEnumerable<Method>? _methods;
-        public IEnumerable<Method> Methods => _methods ??= Models.Select(CreateMethod);
+        // TODO: remove this intermediate state once we generate it before output types
+        private IReadOnlyList<Method>? _methods;
 
-        public IEnumerable<SerializableObjectType> Models { get; }
+        // This method should only be called from OutputMethods as intermediate state.
+        private IReadOnlyList<Method> ShouldNotBeUsedForOutput([CallerMemberName] string caller = "")
+        {
+            Debug.Assert(caller == nameof(OutputMethods) || caller == nameof(SignatureType), $"This method should not be used for output. Caller: {caller}");
+            return _methods ??= _models.Select(CreateMethod).ToList();
+        }
+
+        private IReadOnlyList<Method>? _outputMethods;
+        public IReadOnlyList<Method> OutputMethods
```

Comment:
Could we just use `Methods` as the name here? The `Methods` property on `TypeProvider` just means the methods a writer should write.
In the future we will have others such like `Properties`, `Constructors`, etc...

Reply:
They are not the same thing.
`Methods` represents the methods generated at this run, it might contain duplicated methods from custom code.
`outputMethods` represents the final methods consumed by the writer, which filter out the existing methods from custom code.
There is a TODO to remove this intermediate result once we move this logic before constructing output types.

Reply:
I am saying the same thing - we could keep the most frequently used property to a simpler name (such as `Methods`), and for the "methods generated at this run", we could name it as `GeneratedMethods`. The writer should not use this.

Ultimately, our writer should looks like this:
```
public void Write()
{
foreach (var method in Methods)
{
// write the method
}

foreach (var property in Property)
{
// write the property
}
}
```
It does not change the concept if we name them as `OutputMethods` and `OutputProperties`, just make it simpler.

Reply:
To clarify this further, since `SignatureType` is used recursively for custom, previous GA contract and current generated codes.
So, `SignatureType.Methods` represents methods of all three above, we can't just rename it to `GeneratedMethods`.
And there is a TODO to refine this once we can have the proper set in place before we construct, adding the TODO to `OutputMethods` as well.

Reply:
Never mind, think your intension is just to have Methods as output for the writer. 
Since the intermediate methods is private, we can just rename `OutputMethods` to `Methods`.